### PR TITLE
[JENKINS-40299] Adapt to the new SCM API for detecting change requests

### PR DESCRIPTION
**Pull request by stephenc**
# Description

Adapt to the correct API for detecting change requests, see [JENKINS-40138](https://issues.jenkins-ci.org/browse/JENKINS-40138).

Blue Ocean ticket [JENKINS-40299](https://issues.jenkins-ci.org/browse/JENKINS-40299).

No tests because your existing tests should cover it and if they don't that's your problem, the API has moved and you are broken without this ;-)

# Submitter checklist
- [x] Link to JIRA ticket in description, if appropriate.
- [x] Change is code complete and matches issue description
- [x] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.
- [ ] Ran Acceptance Test Harness against PR changes.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

@reviewbybees 

…quests

### CODE REVIEW

**Comment by stephenc** [comment]
Will also probably need https://github.com/jenkinsci/github-branch-source-plugin/pull/93 before you can test fully (unless you decide to write some tests using the new https://github.com/jenkinsci/scm-api-plugin/tree/master/src/test/java/jenkins/scm/impl/mock mock SCM implementation, e.g. see https://github.com/jenkinsci/branch-api-plugin/blob/12b8364d0fb1aa057b5a4ca81317b150a16121e1/src/test/java/integration/EventsTest.java for some examples of how to write tests using the new mocks

**Comment by michaelneale** [comment]
so this breaks users out in the wild who update to SCM 2.0 - what happened to being backwards compaitble :bug:

**Comment by stephenc** [comment]
Not possible: the attaching of actions to SCMHead is incompatible with event support.

**Comment by michaelneale** [comment]
well it passes the ATH - but we don't test for PR's specifically (as we don't have github enterprise in the ATH yet) in github, but other things are tested for. 

This breaks multibranch in other ways though (as tyler pointed out)

**Comment by michaelneale** [comment]
@stephenc can you be a bit less vague about what is needed to test this? 

even manual instructions. 

Also why was SCM 2.0 released without: https://github.com/jenkinsci/github-branch-source-plugin/pull/93 ?

**Comment by michaelneale** [comment]
Also this needs a blue ocean ticket to track, as I think adding tests using that new mock stuff is a good idea, but more work to do.

**Comment by i386** [comment]
@stephenc I really appreciate that you came through with this patch. Lets time the release for all of this going out.

**Comment by i386** [comment]
@stephenc  🐛  when other components that depend on `scm-api` have been released we will need to bump the versions of those defined in our POMs

**Comment by michaelneale** [comment]
yes, thanks for the change, but my remaining concerns are around other effects with multibranch users (not this PR). This PR passes tests BTW, so could go in any time, just not sure when is right. 

It seems like it should be done when multibranch is updated right? 

I would like to use that mock api too... can we test PRs with it?

**Comment by stephenc** [comment]
@michaelneale https://github.com/jenkinsci/branch-api-plugin/blob/12b8364d0fb1aa057b5a4ca81317b150a16121e1/src/test/java/integration/CategorizationTest.java#L147-L178 shows creating change requests

**Comment by michaelneale** [comment]
@stephenc 💯 on that - most useful. And I like the term "change request" which I assume is generic-speak for pull request (it translates well to non git systems I have encourntered). I am not sure if we can use that in the ATH (which is where we test those tabs) though, but it does help in any case.

**Comment by stephenc** [comment]
@michaelneale if you use "execute groovy script" you can create a controller and set up the controller's repo as you want. then use that from the ath

**Comment by i386** [comment]
Added a link to the Blue Ocean ticket.

**Comment by stephenc** [comment]
Needs https://github.com/jenkinsci/github-branch-source-plugin/pull/93 and https://github.com/jenkinsci/branch-api-plugin/pull/58

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/MultiBranchPipelineImpl.java [hunk 1]
```diff
@@ -264,8 +264,8 @@ private int countJobs(boolean pullRequests) {
         return counter;
     }
     private boolean isPullRequest(Job job) {
-        SCMHead head = SCMHead.HeadByItem.findHead(job);
-        return head != null && head.getAction(ChangeRequestAction.class) != null;
+        // TODO probably want to be using SCMHeadCategory instances to categorize them instead of hard-coding for PRs
```
Comment (i386) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/MultiBranchPipelineImpl.java]: @stephenc that is awesome. We should make Blue Ocean work with that.

Comment (i386) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/MultiBranchPipelineImpl.java]: So would 'Branches' and 'Pull Requests' be different categories provided by Github Branch Source? How do we enumerate categories?

**Comment by michaelneale** [comment]
@stephenc are you able to update your branch again to master - as there were some fixes in recently which the ATH is expecting. 

Otherwise, if we could move this to a branch than some of us could take over if needs be

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/MultiBranchPipelineImpl.java (hunk 1 above)
Comment (stephenc) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/MultiBranchPipelineImpl.java]: @i386 No. The recommended display order is by localized display name and the default category is always present and always has the `getName()` of `default`

**Comment by stephenc** [comment]
@michaelneale I'll be updating this again with the latest plugin versions in the 10 affected plugins, so I'll rebase onto the head of master for that.

Currently looks like we will be doing the release party on Friday: https://github.com/jenkinsci/git-plugin/pull/455#issuecomment-266590777

File: blueocean-commons/pom.xml [hunk 2]
```diff
@@ -5,7 +5,7 @@
   <parent>
     <groupId>io.jenkins.blueocean</groupId>
     <artifactId>blueocean-parent</artifactId>
-    <version>1.0.0-b15-SNAPSHOT</version>
+    <version>1.0.0-beta-15-SNAPSHOT</version>
```
Comment (michaelneale) [inline blueocean-commons/pom.xml]: how come this is change to be the experimental update centre? I guess a temporary thing?

**Comment by i386** [comment]
@stephenc whats the time frame for your changes going out into the main UC?

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/BranchImpl.java [hunk 3]
```diff
@@ -39,12 +41,18 @@ public BranchImpl(Job job, Link parent) {
...
     @Exported(name = PULL_REQUEST, inline = true)
     public PullRequest getPullRequest() {
+        // TODO probably want to be using SCMHeadCategory instances to categorize them instead of hard-coding for PRs
         SCMHead head = SCMHead.HeadByItem.findHead(job);
-        if(head != null) {
-            ChangeRequestAction action = head.getAction(ChangeRequestAction.class);
-            if(action != null){
-                return new PullRequest(action.getId(), action.getURL().toExternalForm(), action.getTitle(), action.getAuthor());
-            }
+        if(head instanceof ChangeRequestSCMHead) {
```
Comment (vivek) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/BranchImpl.java]: LGTM 🐝

**Comment by vivek** [comment]
Changes LGTM 🐝. There is conflict that needs to be resolved, then we can get CI run and ATH pass then  it looks good all good.

**Comment by i386** [comment]
I tried this out locally and received a strange error when indexing my pipeline

```
Started by user admin istrator
[Fri Jan 06 21:11:29 AEDT 2017] Starting branch indexing...
[Fri Jan 06 21:11:30 AEDT 2017] Finished branch indexing. Indexing took 1 sec
FATAL: Failed to recompute children of App Store
java.io.FileNotFoundException: https://api.github.com/repos/i386/app-store-demo
	at com.squareup.okhttp.internal.huc.HttpURLConnectionImpl.getInputStream(HttpURLConnectionImpl.java:243)
	at com.squareup.okhttp.internal.huc.DelegatingHttpsURLConnection.getInputStream(DelegatingHttpsURLConnection.java:210)
	at com.squareup.okhttp.internal.huc.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:25)
	at org.kohsuke.github.Requester.parse(Requester.java:535)
	at org.kohsuke.github.Requester._to(Requester.java:262)
	at org.kohsuke.github.Requester.to(Requester.java:224)
	at org.kohsuke.github.GitHub.getRepository(GitHub.java:369)
	at org.jenkinsci.plugins.github_branch_source.GitHubSCMSource.retrieveActions(GitHubSCMSource.java:967)
	at jenkins.scm.api.SCMSource.fetchActions(SCMSource.java:689)
	at jenkins.branch.MultiBranchProject.computeChildren(MultiBranchProject.java:440)
	at com.cloudbees.hudson.plugins.folder.computed.ComputedFolder.updateChildren(ComputedFolder.java:187)
	at com.cloudbees.hudson.plugins.folder.computed.FolderComputation.run(FolderComputation.java:137)
	at jenkins.branch.MultiBranchProject$BranchIndexing.run(MultiBranchProject.java:786)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:410)
Finished: FAILURE
```

File: blueocean-commons/pom.xml (hunk 2 above)
Comment (michaelneale) [inline blueocean-commons/pom.xml]: yes this can't be beta

**Comment by stephenc** [comment]
Pre-conflict resolution and rebase: https://github.com/jenkinsci/blueocean-plugin/pull/650/commits/a428cd34da42d5af8dc5340d4d06ba58dab63f80

File: blueocean-rest-impl/pom.xml [hunk 4]
```diff
@@ -26,7 +26,7 @@
         <dependency>
             <groupId>org.jenkins-ci.plugins</groupId>
             <artifactId>scm-api</artifactId>
-            <version>2.0</version>
+            <version>2.0.1-beta-1</version>
```
Comment (vivek) [inline blueocean-rest-impl/pom.xml]: Shouldn't this be 2.0.1?

Comment (stephenc) [inline blueocean-rest-impl/pom.xml]: yes

**Comment by vivek** [comment]
@stephenc thanks. I see locally all tests passing. Lets wait for CI and ATH to pass. I am doing some manual testing locally, if all looks good we should be good to go on this.

**Comment by michaelneale** [comment]
@vivek on dog food once this is out the plugins will have to be updated from the update centre there. There is no easy way to really test it unfortunately.

**Comment by vivek** [comment]
@michaelneale right to deploy on dogfood you have to wait till new scm and other affected dependencies land in UC. I was trying to do mvn hpi:run locally on fresh work dir, blueocean is not loading. ATH is failing as tests need to be changed as branch names are name mangled underneath to generate alpha numeric names. This also means we have to stop encoding/decoding branch names.

**Comment by michaelneale** [comment]
@vivek the URIs have changed as part of this?

**Comment by vivek** [comment]
@stephenc blueocean fails to load. Try:

```
mvn clean install
cd blueocean
mvn hpi:run
```

I have captured log at: https://gist.github.com/vivek/576a57f54aafd9d2a42a9bb11ab761c6.I see there is cyclic dependency detected in workflow-scm-step.

```
SEVERE: found cycle in plugin dependencies: (root=Plugin:workflow-scm-step, deactivating all involved) Plugin:workflow-scm-step -> Plugin:git -> Plugin:workflow-scm-step
Jan 16, 2017 1:42:45 PM jenkins.InitReactorRunner$1 onAttained
INFO: Listed all plugins
Jan 16, 2017 1:42:46 PM hudson.init.impl.InstallUncaughtExceptionHandler$DefaultUncaughtExceptionHandler uncaughtException
SEVERE: A thread (pool-6-thread-8/41) died unexpectedly due to an uncaught exception, this may leave your Jenkins in a bad way and is usually indicative of a bug in the code.
java.lang.NoClassDefFoundError: jenkins/plugins/git/AbstractGitSCMSource
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
	at jenkins.util.AntClassLoader.defineClassFromData(AntClassLoader.java:1139)
	at hudson.ClassicPluginStrategy$AntClassLoader2.defineClassFromData(ClassicPluginStrategy.java:878)
	at jenkins.util.AntClassLoader.getClassFromStream(AntClassLoader.java:1310)
	at jenkins.util.AntClassLoader.findClassInComponents(AntClassLoader.java:1366)
	at jenkins.util.AntClassLoader.findClass(AntClassLoader.java:1326)
	at jenkins.util.AntClassLoader.loadClass(AntClassLoader.java:1079)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at java.lang.Class.getDeclaringClass0(Native Method)
	at java.lang.Class.getDeclaringClass(Class.java:1235)
	at java.lang.Class.getEnclosingClass(Class.java:1277)
	at java.lang.Class.getSimpleBinaryName(Class.java:1443)
	at java.lang.Class.getSimpleName(Class.java:1309)
	at hudson.init.TaskMethodFinder.getDisplayNameOf(TaskMethodFinder.java:80)
	at hudson.init.TaskMethodFinder$TaskImpl.getDisplayName(TaskMethodFinder.java:167)
	at jenkins.InitReactorRunner$1.onTaskStarted(InitReactorRunner.java:64)
	at org.jvnet.hudson.reactor.ReactorListener$Aggregator.onTaskStarted(ReactorListener.java:89)
	at org.jvnet.hudson.reactor.Reactor$2.run(Reactor.java:208)
	at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:117)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ClassNotFoundException: jenkins.plugins.git.AbstractGitSCMSource
	at jenkins.util.AntClassLoader.findClassInComponents(AntClassLoader.java:1376)
	at jenkins.util.AntClassLoader.findClass(AntClassLoader.java:1326)
	at jenkins.util.AntClassLoader.loadClass(AntClassLoader.java:1079)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 23 more
```

**Comment by michaelneale** [comment]
replaced by @vivek and my branch of these changes: https://github.com/jenkinsci/blueocean-plugin/pull/715

**Pull request closed without merging**